File: common/src/main/scripts/lib/micropython/oc2/clock.py
import time


def ms():
    return time.ticks_ms()


def deadline(timeout):
    return time.ticks_add(time.ticks_ms(), timeout)


def remaining(deadline):
    left = time.ticks_diff(deadline, time.ticks_ms())
    return left if left > 0 else 0


def expired(deadline):
    return time.ticks_diff(deadline, time.ticks_ms()) <= 0

File: common/src/main/scripts/lib/lua/oc2/clock.lua
local time = require("posix.time")

local clock = {}

function clock.ms()
  local now = time.clock_gettime(time.CLOCK_MONOTONIC)
  return now.tv_sec * 1000 + now.tv_nsec // 1000000
end

function clock.deadline(timeout)
  return clock.ms() + timeout
end

function clock.remaining(deadline)
  local left = deadline - clock.ms()
  return left > 0 and left or 0
end

function clock.expired(deadline)
  return deadline - clock.ms() <= 0
end

return clock
